torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,485 @@
+        for device, per_device in self.per_device_params.items():
+            # Allocate one buffer per rank and per device to group the small parameters
+            self.buckets[device] = [
+                torch.zeros(bucket_cap_kb, dtype=per_device[0][0].dtype, device=device) for _ in range(len(per_device))
+            ]
+        self.should_bucket_param: List[bool] = []
+        self.work_handles: Deque[Any] = deque()
+        self._setup_bucket_strategy()
+        self.initialized = True
+
+    def add_param_group(self, param_group: dict) -> None:
+        """Add a param group to the :class:`Optimizer` s `param_groups`.
+
+        This can be useful when fine tuning a pre-trained network as frozen layers can be made
+        trainable and added to the :class:`Optimizer` as training progresses.
+
+        Arguments:
+            param_group (dict): Specifies what Tensors should be optimized along with group
+            specific optimization options
+
+        .. warning: This handles updating the shards on all partitions, but needs to be called on all ranks.
+        """
+
+        super().add_param_group(param_group)
+        if self.initialized:
+            # Force a re-partitioning
+            self._partition_parameters.clear()
+            self._per_device_params.clear()
+            self._param_rank.clear()
+
+            param_groups = self.partition_parameters()[self.rank]
+            if len(param_groups) == len(self.optim.param_groups) + 1:
+                self.optim.add_param_group(param_groups[-1])
+
+            # Update the bucketing strategy accordingly
+            self._setup_bucket_strategy()
+
+    def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
+        """Update the consolidated state_dict list, one per rank.
+
+        .. warning: This needs to be called on all replicas"""
+
+        # Sync lr and other attributes in case its been updated
+        self._sync_param_groups()
+
+        local_cpu_state = _recursive_copy_to_device(
+            self.local_state_dict(), non_blocking=True, device=torch.device("cpu")
+        )
+
+        # Pull the sharded state from all the other replicas
+        # Store all the states in order, rank by rank
+        if self.rank == recipient_rank:
+            logging.debug("Pulling the sharded optimizer state from all replicas")
+            self._all_states = []
+            for rank in range(self.world_size):
+                if rank == self.rank:
+                    self._all_states.append(local_cpu_state)
+
+                    # Sync with other replicas, almost noop
+                    dist.broadcast_object_list([0], src=self.global_rank, group=self.group)
+                else:
+                    # Fetch the optim state from the other replicas
+                    global_rank = _get_global_rank(self.group, rank)
+                    replica_state = [0]
+                    dist.broadcast_object_list(replica_state, src=global_rank, group=self.group)
```

Comment:
curious, IIUC, this is to collect all states to the `recipient_rank`. If that's the case, any reason for using `broadcast` instead of gather?

Reply:
Not compatible with NCCL ?

Reply:
ahh, yes, sorry, I forgot about that

Reply:
Shall we mention that in the comments, so that future maintainers will know the caveat.

Reply:
I think that a better follow up could be with remote references, so that you don't have to call 'consolidate' beforehand and any rank works, it was the idea behind the distributed data broker that I suggested